zebra-chain/src/primitives/redpallas/batch.rs

```diff
@@ -0,0 +1,265 @@
+    ///
+    /// This is useful (in combination with `Item::clone`) for implementing fallback
+    /// logic when batch verification fails. In contrast to
+    /// [`VerificationKey::verify`](crate::VerificationKey::verify), which requires
+    /// borrowing the message data, the `Item` type is unlinked from the lifetime of
+    /// the message.
+    #[allow(non_snake_case)]
+    pub fn verify_single(self) -> Result<(), Error> {
+        match self.inner {
+            Inner::Binding { vk_bytes, sig, c } => VerificationKey::<Binding>::try_from(vk_bytes)
+                .and_then(|vk| vk.verify_prehashed(&sig, c)),
+            Inner::SpendAuth { vk_bytes, sig, c } => {
+                VerificationKey::<SpendAuth>::try_from(vk_bytes)
+                    .and_then(|vk| vk.verify_prehashed(&sig, c))
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+/// A batch verification context.
+pub struct Verifier {
+    /// Signature data queued for verification.
+    signatures: Vec<Item>,
+}
+
+impl Verifier {
+    /// Construct a new batch verifier.
+    pub fn new() -> Verifier {
+        Verifier::default()
+    }
+
+    /// Queue an Item for verification.
+    pub fn queue<I: Into<Item>>(&mut self, item: I) {
+        self.signatures.push(item.into());
+    }
+
+    /// Perform batch verification, returning `Ok(())` if all signatures were
+    /// valid and `Err` otherwise.
+    ///
+    /// The batch verification equation is:
+    ///
+    /// h_G * -[sum(z_i * s_i)]P_G + sum(\[z_i\]R_i + [z_i * c_i]VK_i) = 0_G
+    ///
+    /// which we split out into:
+    ///
+    /// h_G * -[sum(z_i * s_i)]P_G + sum(\[z_i\]R_i) + sum([z_i * c_i]VK_i) = 0_G
+    ///
+    /// so that we can use multiscalar multiplication speedups.
+    ///
+    /// where for each signature i,
+    /// - VK_i is the verification key;
+    /// - R_i is the signature's R value;
+    /// - s_i is the signature's s value;
+    /// - c_i is the hash of the message and other data;
+    /// - z_i is a random 128-bit Scalar;
+    /// - h_G is the cofactor of the group;
+    /// - P_G is the generator of the subgroup;
+    ///
+    /// Since RedPallas uses different subgroups for different types
+    /// of signatures, SpendAuth's and Binding's, we need to have yet
+    /// another point and associated scalar accumulator for all the
+    /// signatures of each type in our batch, but we can still
+    /// amortize computation nicely in one multiscalar multiplication:
+    ///
+    /// h_G * ( [-sum(z_i * s_i): i_type == SpendAuth]P_SpendAuth + [-sum(z_i * s_i): i_type == Binding]P_Binding + sum(\[z_i\]R_i) + sum([z_i * c_i]VK_i) ) = 0_G
+    ///
+    /// As follows elliptic curve scalar multiplication convention,
+    /// scalar variables are lowercase and group point variables
+    /// are uppercase. This does not exactly match the RedDSA
+    /// notation in the [protocol specification §B.1][ps].
+    ///
+    /// [ps]: https://zips.z.cash/protocol/protocol.pdf#reddsabatchverify
```

Comment:
That's a neat optimization! Out of curiosity, where is this from? I know that it's specified in the spec, but the spec does not seem to refer to anything else. It may be useful for reader to understand e.g. why the random scalars are needed, why 128 bits are enough, etc. (I think I've deduced the answers but I'm curious to check them :laughing: )

Reply:
It's from [[BDLSY2012](https://cr.yp.to/papers.html#ed25519] (the main Ed25519 paper).

Reply:
Thanks @daira!

Reply:
👍